### DISCUSSION TITLE
Adding new category to multi-label classification problem

### DISCUSSION THREAD

**Discussion by drisspg** [explosion/spaCy#8171] [Help: Best practices]
## Scenario: 
I have generated dataset X<sub>0</sub> for multi-label classification that has N categories.  I train model everything looks good and then product comes back and says we want another category. I say okay I generate dataset X<sub>1</sub> with N+1 categories.  X<sub>0</sub> and  X<sub>1</sub> are disjoint in this scenario. 

## Problem:
It is expensive to generate annotated datasets. Instead of relabel  X<sub>0</sub> and only add the extra category when relevant,  X<sub>1</sub> is a separate text dataset that has been labeled to try and improve existing performance on the original N categories and provide labeled trainng data for the new N+1 category. If a new dataset was made X = X<sub>0</sub> U X<sub>1</sub> the problem arrises that in the original  X<sub>0</sub> dataset some examples should have been labeled with the N+1 category.  Training a model on X may not result in good performance because a similar instance in X<sub>0</sub> did not get labeled with the N+1 category and it did in X<sub>1</sub>. I think this would likely hurt the recall of the new category.

## Potential solutions

1. Create new dataset X =  X<sub>0</sub> U {X<sub>1</sub> | N+1 category is not present} and Y = {X<sub>1</sub> | N+1 category is present}. Train existing model on dataset X and create new model on dataset Y. I have updated to spacy 3.0 and already have a pipeline with 2 text cats ( merge two separately trained into one model). The main limitation of this is that we lose any nice correlations that may be present for the previous N categories and the new N+1 category.

2. Create new dataset X = X<sub>0</sub> U X<sub>1</sub> and retrain model. Accept the potential loss in performance do to conflicting instance labels but hopefully as X<sub>1</sub> grows in size the loss in performance will decrease. 

3. Train model on X<sub>1</sub>. Use this new model to infer on all of X<sub>0</sub>. If prediction for an instance in X<sub>0</sub> is significant for the N+1 category, amend that instance label with this new N+1 category being present. After this has been done X = X<sub>0</sub> U X<sub>1</sub> and retrain model with the hope that this beat solution 2. 

Has anyone had a similar experience where their multi-label classification problem is learned continually and has new labels that potentially overlap with previous datasets. I am also in a similar scenario for NER.

**Comment by svlandeg** [comment]
Hi!

I just wanted to add one more option for the future to consider. We're currently [working on](https://github.com/explosion/spaCy/pull/7862) implementing a resizable `textcat`, which means that you'll be able to add new labels to an already trained `textcat` model. I think for the annotations that have changed label, you'll want to make sure that you retrain/fine-tune further on those, and the model will likely adjust itself accordingly.
